### ISSUE TITLE
error in Mamdani 

### ISSUE THREAD

**Issue by RakshaAg** [aresio/simpful#8]
I am getting the following error possibly due to use of AND in the rules.

* Indices: {'High': 2, 'Low': 0, 'Average': 1}
 * Weighted values: 0.00	Values: 0.00	CoG: nan
{'Rank': nan}
/usr/local/lib/python3.7/dist-packages/simpful/simpful.py:510: RuntimeWarning: invalid value encountered in double_scalars
  CoG = sumwv/sumv

Kindly help

**Comment by aresio** [comment]
Dear anonyous297,

could you please post a minimal snippet of code that generates that error? That would help us to understand what is going wrong.

**Comment by FlowRiderGit** [comment]
I might be able to shed some light on this, since I encountered a similar issue a few days ago.

### TL;DR: Currently the Mamdani inference only works, if there is only one rule per consequent term.

I have been using a system with three input variables and one output variables and created a complete AND connected rule base for this system. Naturally, this led to multiple rules with different premises, but the same consequent term.
When executing Mamdani inference, simpful however currently only takes one rule per consequent term into account.
To be exact, in the case of multiple rules with the same consequent term (as in my case) only the last rule per term is taken into account.
This is due to the code below from the mediate_Mamdani() function in the simpful.py file:



		cuts_list = defaultdict()

		x0, x1 = self._lvs[output].get_universe_of_discourse()

		for (ant, res) in zip(antecedent, results):

			outname = res[0]
			outterm = res[1]

			if verbose:	
				print(" ** Rule composition:", ant, "->", res, ", output variable: '%s'" % outname, "with term: '%s'" % outterm)			

			if outname==output:

				try:
					value = ant.evaluate(self) 
				except RuntimeError: 
					raise Exception("ERROR: one rule could not be evaluated\n"
					+ " --- PROBLEMATIC RULE:\n"
					+ "IF " + str(ant) + " THEN " + str(res) + "\n")

				cuts_list[outterm] = value


The problem here is that cuts_list only has a single value for each output term and the saved values get overwritten for each new rule.
Therefore, if by accident the evaluated rule firing strength of each **last** rule per consequent term is 0, the values of sumwv and sumv will also be 0 and one gets the incorrect result that anonymous297 showed: 

> Weighted values: 0.00 Values: 0.00 CoG: nan

I was able to prove this behaviour by removing rules from the end of my rule base until the last rule in the rule base had a firing strength >0, in which case CoG had a numeric value.

### Here is the fix I used:
1. Change the type of the cuts_list defaultdict to list
2. Append values to the cuts_list outterms instead of replacing them
3. Add a loop over the list of values in each cuts_list outterm for calculating the "result" variable



		final_result = {}

		for output in outputs:

			if verbose:
				print(" * Processing output for variable '%s'" %  output)
				print("   whose universe of discourse is:", self._lvs[output].get_universe_of_discourse())
				print("   contains the following fuzzy sets:", self._lvs[output]._FSlist )
			cuts_list = defaultdict(list)

			x0, x1 = self._lvs[output].get_universe_of_discourse()

			for (ant, res) in zip(antecedent, results):

				outname = res[0]
				outterm = res[1]

				if verbose:	
					print(" ** Rule composition:", ant, "->", res, ", output variable: '%s'" % outname, "with term: '%s'" % outterm)			

				if outname==output:

					try:
						value = ant.evaluate(self)
					except RuntimeError: 
						raise Exception("ERROR: one rule could not be evaluated\n"
						+ " --- PROBLEMATIC RULE:\n"
						+ "IF " + str(ant) + " THEN " + str(res) + "\n")

					cuts_list[outterm].append(value)

			values = []
			weightedvalues = []
			integration_points = linspace(x0, x1, subdivisions)

			convenience_dict = {}
			for k in cuts_list.keys():
				convenience_dict[k] = self._lvs[output].get_index(k)
			if verbose: print ( " * Indices:", convenience_dict)

			for u in integration_points:
				#print ("x=%.1f" % u)
				comp_values = []
				for k,vl in cuts_list.items():
					# result = float(self._outputfuzzysets[k].get_value_cut(u, cut=v))
					for v in vl:
						n = convenience_dict[k]					
						fs_term = self._lvs[output]._FSlist[n]
						result = float(fs_term.get_value_cut(u, cut=v))
						comp_values.append(result)
				keep = max(comp_values)
				values.append(keep)
				weightedvalues.append(keep*u)

			sumwv = sum(weightedvalues); sumv = sum(values)
			CoG = sumwv/sumv


And thats it. Now each rule will get considered and the CoG value will only give nan, if no rule fires at all, in which case a rule is missing from the rule base.

**Comment by aresio** [comment]
Thank you very much indeed FlowRiderGit for your help! We will study your code and integrate your patch asap into Simpful.

**Comment by sspola** [comment]
Dear @anonymous297 and @FlowRiderGit thanks for pointing out the problem.
I pushed a new version (d03377f9a6d2da2acc8435b671ff7e2f2c8f3727), implementing the fix suggested by @FlowRiderGit.

I hope everything works well now!
